services/galley/src/Galley/API/Util.hs

```diff
@@ -759,25 +754,78 @@ fromConversationCreated loc rc@ConversationCreated {..} =
         ProtocolProteus
 
 -- | Notify remote users of being added to a new conversation. The return value
--- consists of users that could not be notified; they will not be considered to
--- be conversation members.
+-- consists of a split of users that could not be notified. Users that could not
+-- be notified will not be considered to be conversation members.
 registerRemoteConversationMemberships ::
   (Member FederatorAccess r) =>
   -- | The time stamp when the conversation was created
   UTCTime ->
-  -- | The domain of the user that created the conversation
-  Domain ->
-  Data.Conversation ->
-  Sem r (Set (Remote UserId))
-registerRemoteConversationMemberships now localDomain c = do
-  let allRemoteMembers = nubOrd (map rmId (Data.convRemoteMembers c))
-      rc = toConversationCreated now localDomain c
-  fmap toSet $ runFederatedConcurrentlyEither allRemoteMembers $ \_ ->
-    fedClient @'Galley @"on-conversation-created" rc
+  Local Data.Conversation ->
+  Sem r DataTypes.MemberAddStatus
+registerRemoteConversationMemberships now lc = do
+  let c = tUnqualified lc
+      allRemoteMembers = nubOrd (Data.convRemoteMembers c)
+      allRemoteMembersQualified = remoteMemberQualify <$> allRemoteMembers
+      allRemoteBuckets :: [Remote [RemoteMember]] = bucketRemote allRemoteMembersQualified
+      rc = toConversationCreated now c
+  failedToNotify <- fmap (foldMap (either (sequenceA . fst) mempty)) $
```

Comment:
this looks weird.

the type of the second argument of this `fmap` is `[Either (Remote [RemoteMember], FederationError) (Remote ())]`.

- shouldn't we keep the buckets with the different remotes, rather than filtering out again below in line 786?
- why does the `FederationError` have to be the same value for all remote members?  shouldn't it be allowed to differ between members?

Reply:
The comment is referring to old code so I'm not sure my reply will address your concerns, but here's an attempt:

1. Regarding keeping them in buckets, there are a few variations of the same data, but in different types, each suitable for a slightly different need. We do have buckets around too.
2. The same `FederationError` question is, I believe, addressed by the observation I made yesterday during the call: there's no incremental update from a remote backend on the progress of processing member updates, so either all of members are added successfully or there is a failure, e.g., a `FederationError` is returned, in which case none of the remote backend's users are considered to be conversation members.